include_core/OMR/TypeTraits.hpp

```diff
@@ -0,0 +1,140 @@
+/*******************************************************************************
+ *  Copyright (c) 2018, 2018 IBM and others
+ *
+ *  This program and the accompanying materials are made available under
+ *  the terms of the Eclipse Public License 2.0 which accompanies this
+ *  distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ *  or the Apache License, Version 2.0 which accompanies this distribution and
+ *  is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ *  This Source Code may also be made available under the following
+ *  Secondary Licenses when the conditions for such availability set
+ *  forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ *  General Public License, version 2 with the GNU Classpath
+ *  Exception [1] and GNU General Public License, version 2 with the
+ *  OpenJDK Assembly Exception [2].
+ *
+ *  [1] https://www.gnu.org/software/classpath/license.html
+ *  [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ *  SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#if !defined(OMR_TYPETRAITS_HPP_)
+#define OMR_TYPETRAITS_HPP_
+
+namespace OMR
+{
+
+///
+/// Basic constants
+///
+
+/// Define an integral constant.
+template <typename T, T value>
+struct IntegralConstant
+{
+	typedef T ValueType;
+
+	static const ValueType VALUE = value;
+
+	ValueType operator()() const { return VALUE; }
+};
+
+template <typename T, T value>
+const T IntegralConstant<T, value>::VALUE;
+
+template <bool value>
+struct BoolConstant : IntegralConstant<bool, value> {};
+
+struct FalseConstant : BoolConstant<false> {};
+
+struct TrueConstant : BoolConstant<true> {};
+
+template <typename T>
+struct TypeAlias
+{
+	typedef T Type;
+};
+
+///
+/// Type modifications: add or remove qualifiers or type modifiers
+///
+
+/// RemoveConst<T> is T without any const qualifiers.
+template <typename T>
+struct RemoveConst : TypeAlias<T> {};
+
+template <typename T>
+struct RemoveConst<const T> : TypeAlias<T> {};
+
+/// RemoveVolatile<T>::Type is T without volatile qualifiers.
+template <typename T>
+struct RemoveVolatile : TypeAlias<T> {};
+
+template <typename T>
+struct RemoveVolatile<volatile T> : TypeAlias<T> {};
+
+/// RemoveCv<T>::Type is T without const/volatile qualifiers.
+template <typename T>
+struct RemoveCv : RemoveVolatile<typename RemoveConst<T>::Type> {};
+
+/// RemoveReference<T>::Type is T as a non-reference.
+template <typename T>
+struct RemoveReference : TypeAlias<T> {};
+
+template <typename T>
+struct RemoveReference<T&> : RemoveReference<T> {};
```

Comment:
Is there a reason for defining the template recursively, instead of:

```suggestion
struct RemoveReference<T&> : TypeAlias<T> {};
```

Unless there's something I'm missing, the two should be equivalent because reference collapsing prevents having a reference to a reference.

Reply:
It was just a stylistic choice. In my initial PR, most types were defined recursively. It makes less sense now.